ssl/statem/extensions_clnt.c

```diff
@@ -198,22 +198,33 @@ EXT_RETURN tls_construct_ctos_supported_groups(SSL *s, WPACKET *pkt,
         uint16_t ctmp = pgroups[i];
 
         if (tls_curve_allowed(s, ctmp, SSL_SECOP_CURVE_SUPPORTED)) {
-            if (!WPACKET_put_bytes_u16(pkt, ctmp)) {
-                    SSLfatal(s, SSL_AD_INTERNAL_ERROR,
-                             SSL_F_TLS_CONSTRUCT_CTOS_SUPPORTED_GROUPS,
-                             ERR_R_INTERNAL_ERROR);
-                    return EXT_RETURN_FAIL;
-                }
+#ifndef OPENSSL_NO_TLS1_3
+            int ctmp13 = ssl_group_id_internal_to_tls13(ctmp);
+            if (ctmp13 != 0 && ctmp13 != ctmp) {
+                int ver_min, ver_max;
+                if (ssl_get_min_max_version (s, &ver_min, &ver_max, NULL))
+                    goto err;
+                if (ver_max >= TLS1_3_VERSION
+                        && !WPACKET_put_bytes_u16(pkt, ctmp13))
+                    goto err;
+                if (ver_min >= TLS1_3_VERSION)
+                    continue;
+            }
+#endif
+            if (!WPACKET_put_bytes_u16(pkt, ctmp))
+                goto err;
         }
     }
-    if (!WPACKET_close(pkt) || !WPACKET_close(pkt)) {
-        SSLfatal(s, SSL_AD_INTERNAL_ERROR,
-                 SSL_F_TLS_CONSTRUCT_CTOS_SUPPORTED_GROUPS,
-                 ERR_R_INTERNAL_ERROR);
-        return EXT_RETURN_FAIL;
-    }
+    if (!WPACKET_close(pkt) || !WPACKET_close(pkt))
+        goto err;
 
     return EXT_RETURN_SENT;
+
+ err:
+    SSLfatal(s, SSL_AD_INTERNAL_ERROR,
```

Comment:
I would rather have  the `SSLfatal` call replicated on each location an error occurs. This means we actually will know the exact line that the error occurred on - rather than "somewhere in this function".

Reply:
Okay.  Good point.